internal/api/v1/middleware/recovery.go

```diff
@@ -0,0 +1,39 @@
+// Copyright © 2021 - 2023 SUSE LLC
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//     http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package middleware
+
+import (
+	"fmt"
+
+	"github.com/epinio/epinio/internal/api/v1/response"
+	"github.com/epinio/epinio/internal/cli/server/requestctx"
+	apierrors "github.com/epinio/epinio/pkg/api/core/v1/errors"
+	"github.com/gin-gonic/gin"
+)
+
+func Recovery() gin.HandlerFunc {
+	return gin.CustomRecovery(func(c *gin.Context, anyerr any) {
+		ctx := c.Request.Context()
+		logger := requestctx.Logger(ctx).WithName("RecoveryMiddleware")
+
+		err, ok := anyerr.(error)
+		if !ok {
+			err = fmt.Errorf("unknown error type occurred [%T]", anyerr)
+		}
+
+		logger.Error(err, "recovered from panic")
+
+		// we don't want to expose internal details to the client
+		response.Error(c, apierrors.NewInternalError("something bad happened"))
+		c.Abort()
```

Comment:
In one of my personal (CGI) apps I place a random hex string into the error message for the user. The same string goes into the app log (for us, server log). That way the user can give the hex string to the operator and the operator is then able to quickly find the place in the logs with the same string and detailed error information.
__Note:__ This works only if the log level of logger above is low enough for the message to be seen even when the server runs at the standard log level.
The user error should mention that the hex string is for the operator.

Reply:
It makes sense! I made the changes. I had to use a different Gin Recovery handler to write the stack to the logs. Fortunately these errors are logged at any level, and also we already had the request ID available.

Pushed.